iceoryx_hoofs/posix/filesystem/source/file.cpp

```diff
@@ -0,0 +1,438 @@
+// Copyright (c) 2023 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iox/file.hpp"
+#include "iceoryx_hoofs/posix_wrapper/posix_call.hpp"
+#include "iceoryx_platform/attributes.hpp"
+#include "iceoryx_platform/errno.hpp"
+#include "iceoryx_platform/fcntl.hpp"
+#include "iceoryx_platform/stdio.hpp"
+#include "iox/filesystem.hpp"
+
+namespace iox
+{
+expected<File, FileCreationError> FileBuilder::create(const FilePath& name) noexcept
+{
+    switch (m_open_mode)
+    {
+    case posix::OpenMode::OPEN_EXISTING:
+        return this->open(name);
+    case posix::OpenMode::PURGE_AND_CREATE:
+        if (File::remove(name).has_error())
+        {
+            IOX_LOG(ERROR) << "Unable to purge and open file \"" << name.as_string()
+                           << "\" since the file could not be removed";
+            return iox::error<FileCreationError>(FileCreationError::CannotBePurged);
+        }
+        IOX_FALLTHROUGH;
+    case posix::OpenMode::EXCLUSIVE_CREATE:
+        return this->open_impl(true, name);
+    case posix::OpenMode::OPEN_OR_CREATE:
+        auto result = this->open_impl(false, name);
+        if (!result.has_error() || result.get_error() != FileCreationError::DoesNotExist)
+        {
+            return result;
+        }
+
+        return this->open_impl(true, name);
```

Comment:
I don't understand why `open_impl` is called twice? Just to get the `IOX_LOG(ERROR) << "Unable to open file since the file does not exist.";` log message?

Reply:
Also the error handling is much easier know and the `open_impl` could be removed. When `O_CREAT` is just set, which is happening under the hood, the file already exists warning does not pop up since it is simply opened.
Only when exclusive open `O_CREAT | O_EXCL` is set then it will result in an error.